crystal_diffusion/models/score_network.py

```diff
@@ -199,3 +215,112 @@ def _forward_unchecked(self, batch: Dict[AnyStr, torch.Tensor]) -> torch.Tensor:
+
+
+class MACEScoreNetwork(ScoreNetwork):
+    """Score network using atom features from MACE.
+
+    Inherits from the given framework's model class.
+    """
+
+    def __init__(self, hyper_params: MACEScoreNetworkParameters):
+        """__init__.
+
+        Args:
+            hyper_params (dict): hyper parameters from the config file.
+        """
+        super(MACEScoreNetwork, self).__init__(hyper_params)
+
+        self.z_table = get_atomic_number_table_from_zs([14])  # TODO more than 1 type & duplicate from dataloader
+        self.r_max = hyper_params.r_max
+        self.collate_fn = Collater(follow_batch=[None], exclude_keys=[None])
+
+        mace_config = dict(
+            r_max=hyper_params.r_max,
+            num_bessel=hyper_params.num_bessel,
+            num_polynomial_cutoff=hyper_params.num_polynomial_cutoff,
+            max_ell=hyper_params.max_ell,
+            interaction_cls=interaction_classes[hyper_params.interaction_cls],
+            interaction_cls_first=interaction_classes[hyper_params.interaction_cls_first],
+            num_interactions=hyper_params.num_interactions,
+            num_elements=len(self.z_table),
+            hidden_irreps=o3.Irreps(hyper_params.hidden_irreps),
+            MLP_irreps=o3.Irreps(hyper_params.MLP_irreps),
+            atomic_energies=hyper_params.atomic_energies,
+            avg_num_neighbors=hyper_params.avg_num_neighbors,
+            atomic_numbers=self.z_table.zs,
+            correlation=hyper_params.correlation,
+            gate=gate_dict[hyper_params.gate],
+            radial_MLP=ast.literal_eval(hyper_params.radial_MLP),
+            radial_type=hyper_params.radial_type
+        )
+
+        self._natoms = hyper_params.number_of_atoms
+
+        self.mace_network = MACE(**mace_config)
+
+        hidden_dimensions = [hyper_params.hidden_dimensions_size] * hyper_params.n_hidden_dimensions
+        self.mace_output_size = 640  # TODO check where 640 - mace features size - comes from
+        self.mlp_layers = nn.Sequential()
+        input_dimensions = [self.mace_output_size] + hidden_dimensions
+        output_dimensions = hidden_dimensions + [self.spatial_dimension]
+        add_relus = len(input_dimensions) * [True]
+        add_relus[-1] = False
+
+        for input_dimension, output_dimension, add_relu in zip(input_dimensions, output_dimensions, add_relus):
+            self.mlp_layers.append(nn.Linear(input_dimension, output_dimension))
+            if add_relu:
+                self.mlp_layers.append(nn.ReLU())
+
+    def _check_batch(self, batch: Dict[AnyStr, torch.Tensor]):
+        super(MACEScoreNetwork, self)._check_batch(batch)
+        number_of_atoms = batch[self.position_key].shape[1]
+        assert (
+            number_of_atoms == self._natoms
+        ), "The dimension corresponding to the number of atoms is not consistent with the configuration."
+
+    def _forward_unchecked(self, batch: Dict[AnyStr, torch.Tensor]) -> torch.Tensor:
```

Comment:
The score network needs the diffusion time as input, ${\bf s}_\theta({\bf x}, t)$.

From what I can see, time is not taken into account in the architecture: `input_to_mace` does not currently set it as a node attribute, and the head MLPs do not use time either.